crates/runtime/src/instance/allocator/pooling/memory_pool.rs

```diff
@@ -393,10 +550,150 @@ impl Drop for MemoryPool {
     }
 }
 
+/// The index of a memory allocation within an `InstanceAllocator`.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub struct StripedAllocationIndex(u32);
+
+impl StripedAllocationIndex {
+    fn from_unstriped_slot_index(index: MemoryAllocationIndex, num_stripes: usize) -> Self {
+        let num_stripes: u32 = num_stripes.try_into().unwrap();
+        Self(index.0 / num_stripes)
+    }
+
+    fn as_unstriped_slot_index(self, stripe: usize, num_stripes: usize) -> MemoryAllocationIndex {
+        let num_stripes: u32 = num_stripes.try_into().unwrap();
+        let stripe: u32 = stripe.try_into().unwrap();
+        MemoryAllocationIndex(self.0 * num_stripes + stripe)
+    }
+}
+
+#[derive(Clone, Debug)]
+struct SlabConstraints {
+    max_memory_bytes: usize,
+    num_memory_slots: usize,
+    num_pkeys_available: usize,
+    guard_bytes: usize,
+    guard_before_slots: bool,
+}
+
+#[derive(Debug)]
+struct SlabLayout {
+    /// The total number of bytes to allocate for the memory pool slab.
+    total_slab_bytes: usize,
+    /// If necessary, the number of bytes to reserve as a guard region at the
+    /// beginning of the slab.
+    pre_slab_guard_bytes: usize,
+    /// Like `pre_slab_guard_bytes`, but at the end of the slab.
+    post_slab_guard_bytes: usize,
+    /// The size of each slot in the memory pool; this comprehends the maximum
```

Comment:
"Comprehends" reads a little weird here?

```suggestion
    /// The size of each slot in the memory pool; this contains the maximum
```